File: Leetcode-solution-python/optimalPartitionOfString.py
def index(c):
    return ord(c) - ord('a')


class Solution:
    def partitionString(self, s: str) -> int:
        n = len(s)
        freq = [False] * 26
        i, j, count = 0, 0, 1
        while j < n:
            if freq[index(s[j])]:
                while i < j:
                    freq[index(s[i])] = False
                    i += 1
                count += 1
            freq[index(s[j])] = True
            j += 1
        return count

File: Leetcode-solution-c-c++/optimalPartitionOfString.cpp
#include<bits/stdc++.h>

#define printArr(arr)                     \
   for (int i = 0; i < arr.size(); i++)   \
       cout << arr[i] << " ";           \
   cout << endl;
#define print2DArr(arr)                        \
   for (int i = 0; i < arr.size(); i++)        \
   {                                           \
       for (int j = 0; j < arr[0].size(); j++) \
       {                                       \
           cout << arr[i][j] << " ";         \
       }                                       \
       cout << endl;                           \
   }

using namespace std;

class Solution {
public:
    int partitionString(string s) {
        int count = 1;
        unordered_set<char> st;
        int n = s.size();
        for(int i=0; i<n; i++) {
            if(st.find(s[i]) == st.end()){
                st.insert(s[i]);
            }else{
                count++;
                st.clear();
                st.insert(s[i]);
            }
        }
        return count;
    }
};